python/tests/io/test_base.py

```diff
@@ -0,0 +1,188 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import os
+import tempfile
+
+import pytest
+
+from iceberg.io.base import FileIO, InputFile, OutputFile
+
+
+class LocalInputFile(InputFile):
+    """An InputFile implementation for local files (for test use only)"""
+
+    def __init__(self, location: str):
+        if not location.startswith("file://"):
+            raise ValueError("LocalInputFile location must start with `file://`")
+        super().__init__(location=location.split("file://")[1])
+
+    def __len__(self):
+        return os.path.getsize(self.location)
+
+    def exists(self):
+        return os.path.exists(self.location)
+
+    def open(self):
+        return open(self.location, "rb")
+
+
+class LocalOutputFile(OutputFile):
+    """An OutputFile implementation for local files (for test use only)"""
+
+    def __init__(self, location: str):
+        if not location.startswith("file://"):
+            raise ValueError("LocalOutputFile location must start with `file://`")
+        super().__init__(location=location.split("file://")[1])
+
+    def __len__(self):
+        return len(self._file_obj)
+
+    def exists(self):
+        return os.path.exists(self.location)
+
+    def to_input_file(self):
+        return LocalInputFile(location=f"file://{self.location}")
+
+    def create(self, overwrite: bool = False) -> None:
+        if not overwrite and self.exists():
+            raise FileExistsError(f"{self.location} already exists")
```

Comment:
Instead of checking for existence directly, I think this should use mode `wbx` when not overwriting, which will [fail if the file already exists](https://docs.python.org/3/library/functions.html#open). That ensures that the check is atomic. With the check here, there is a race condition between two writers that are in this method. Both check that the file doesn't exist and succeed, but then both try to create the file.

Reply:
Awesome, looks like 'xb' is the right mode description so I updated this and it looks much cleaner!:
```py
    def create(self, overwrite: bool = False) -> None:
        return open(self.parsed_location.path, "wb" if overwrite else "xb")
```
relevant commit: [25836bf](https://github.com/apache/iceberg/pull/3691/commits/25836bf0c8aa546d913b67faba323265498628d6)

Reply:
Do you get the correct `FileExistsError` from `open`?

Reply:
Yep! I validate that [here](https://github.com/apache/iceberg/pull/3691/files#diff-3ced008699fedda78712dccd78a398136c8a279342fde9ac9c1713ddfd891339R167) in one of the tests.